Task/raja 306 forall indexset tests

rhornung67 opened the pull request:
# Summary

- This PR adds tests in the new test organization for forall methods using index sets.
- Perhaps, more importantly, it moves us closer to being able to specify at run-time a memory resource for list segment index data to live. Prior to this, the choice was made at compile time depending on whether CUDA or HIP was enabled and the internal implementation used unified memory for both.
- Note that I left all the machinery for the original list segment memory management in place so that the original behavior is maintained, in addition to the new behavior which is used in the new forall list segment and forall index set tests. Eventually, the old machinery can be removed.
- An important goal of this PR is to decide how we want to do the list segment memory management going forward, including moving users to the new memory management model. I would like to move to a model where the TypedListSegment class is also templated on a camp resource. I didn't do that here to preserve the original interface of the class.

trws commented:
I started a custom force build on travis [here](https://travis-ci.org/github/LLNL/RAJA/builds/670821032) so I can see what's going on, it may or may not show up as a check here but that's what's up if something happens.

willkill07 commented on include/RAJA/index/ListSegment.hpp:
```diff
@@ -280,34 +387,68 @@ class TypedListSegment
   // owns the index data.
   //
-  void initIndexData(const value_type* container,
+  void initIndexData(bool use_resource,
```
Perhaps this should be protected/private since the user shouldn't be calling this

rhornung67 replied:
It is private.

willkill07 commented on include/RAJA/index/ListSegment.hpp:
```diff
@@ -189,27 +281,32 @@ class TypedListSegment
-    allocate_and_copy<Has_GPU>(container);
-    m_owned = Owned;
+    if (m_size > 0) {
+      allocate_and_copy<Has_GPU>(container);
```
I believe this is still an artifact -- should be using initIndexData here

rhornung67 replied:
I'm going to leave this as is for now. It will go away in the future.

willkill07 commented on include/RAJA/index/ListSegment.hpp:
```diff
@@ -189,27 +281,32 @@ class TypedListSegment
   /// Copy-constructor for list segment.
   ///
   TypedListSegment(const TypedListSegment& other)
+    : m_resource(other.m_resource), m_use_resource(other.m_use_resource)
```
I know this is pedantic, but providing a default value for all members would be useful

```cpp
, m_owned (Unowned), m_data (nullptr), m_size(0)
```

rhornung67 replied:
Done!

willkill07 commented on include/RAJA/index/ListSegment.hpp:
```diff
@@ -165,20 +175,102 @@ class TypedListSegment
                    Index_type length,
                    IndexOwnership owned = Owned)
+    : m_resource(camp::resources::Resource{camp::resources::Host()}),
+      m_use_resource(false)
```
Should default init all members where possible

rhornung67 replied:
Done!

rhornung67 commented:
@willkill07 thanks for the feedback. I agree with all your comments. This PR was intended to get the ball rolling for how I think we should eventually move forward. The code is a bit messy and I left in all the old stuff so our tests that have not been converted to the new organization will still work and this will not surprise users. I think the best approach is to note future changes in the next RAJA release to give users a heads up on what's coming (i.e., requiring them to specify a camp resource for each TypedListSegment they create), and then force the new interface in the subsequent release.

rhornung67 commented:
@trws it looks like the CI is still running on travis-ci.org. Will this be switched over to travis-ci.com?  Also, the failing tests are expected. We have issues with range-stride segments and the TBB back-end. Apparently, we have never tried that combination before, at least in our tests. @mdavis36 is digging into it.

trws commented:
He emailed me about it.

As to the transition, I’m waiting to hear back from other Llnl GitHub org admins before subscribing the org to the beta, assuming that happens I’ll transition the raja repos shortly thereafter. For now I just got this one unstuck on the existing webhook setup.

trws commented:
This branch should be rather happier now.  The warnings are fixed in an updated camp that is referenced in the commit I added, as are all the warnings from the numeric_iterators.  The way I fixed the iterators may cause issues with HIP or CUDA versions I don’t have locally, but should be easily fixed if Travis pops something back.

trws commented:
Well, the good news is this was even more useful than it might have been.  The bad news is I found a latent bug in our usage of std::distance that is going to be annoying to patch.  I'm adding that one to my list to fix imminently but not planning to fix it here.  Summary is that `distance` can return a negative value in some cases after c++11, and we assume it can't.

trws commented:
I found the remaining problem here, but want some opinions on policy before I fix it.

@ajkunen, @mdavis36, @rhornung67 and @willkill07, you've all touched right around this code so I'd like your thoughts if possible.  I fixed the issue in the tbb backend that was causing it to reject the negative stride ranges out of hand.  WIth that fixed, it treated them as ranges in a classic C++ for loop using the != check, and then corrupted memory and caused a memunmap failure and abort.  This is because the two tests using ranges like this: `RangeStrideSegment(0, 20, -2)` create iterators that are technically valid as far as `std::distance` and the loop and all of that are concerned, but generate insanely huge ranges across negative offsets.  In other words, it generates an invalid RangeStrideSegment in that it doesn't give out valid RandomAccessIterator values for `begin()` and `end()`.

Options to fix this:

1. Throw an exception or abort if the user gives us invalid values, probably not but worth mentioning
2. If the end is not reachable from the begin without wrapping, clamp the begin and end iterators such that `begin()==end()` is true in addition to clamping size to 0 (we already clamp the size)
3. Fake it in the tbb backend by either checking the size or using a different loop type to avoid the problem but leave it there

I'm leaning toward 2, but it could leave latent bugs around by making invalid ranges just not execute anything, so I'd prefer to hear other input before doing it.

willkill07 commented:
I agree with option 2 the most. It would be very easy to detect 

```cpp
(stride > 0)
? std::distance(begin, end) >= 0
: std::distance(begin, end) <= 0
```

mdavis36 commented:
I think I'm leaning towards option 2 as well. I believe that currently when invalid ranges are used with any of the segment types they don't don't execute anything.

trws commented:
That’s mostly true, but if you give a `TypedRangeSegment` an `end` 
that’s below its `begin` it will throw an exception or abort.  I 
realize I forgot to say so yesterday, but that’s a non-trivial part of 
why I’m asking, we use both options 1 and 2 currently.

On 6 Apr 2020, at 7:53, Michael Davis wrote:

> I think I'm leaning towards option 2 as well. I believe that currently 
> when invalid ranges are used with any of the segment types they don't 
> don't execute anything.
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/LLNL/RAJA/pull/790#issuecomment-609843884

rhornung67 commented:
@trws It's not clear to me what you mean in #1 by "invalid values". Do you mean when the values given in a range or range-stride segment would result in no loop iterations to execute? To me it makes most sense to follow the semantics of standard C-style for loops. For example,
'for (int i = 2; i < 0; ++i)' would yield no iterations to execute, but the code would still compile and not generate any errors when run (except for unexpected behavior if the user screwed up). However, 'for (int i = 6; i > -1: i -= 2)' would yield iterations with values i = 6, 4, 2, 0. This would be the same as using TypedRangeSegment(6, -1, -2) with a RAJA forall, which we need to support.  Perhaps, we could add run time bounds checks inserted via our compile-time CMake variable internally, like we do for Views, and report to users that no loop iterations would result?

Thanks for digging into this.

rhornung67 commented on include/RAJA/policy/tbb/forall.hpp:
```diff
@@ -77,15 +73,17 @@ RAJA_INLINE void forall_impl(const tbb_for_dynamic& p,
                         auto body = privatizer.get_priv();
-                        for (const auto& i : r)
-                          body(i);
+                        for (auto i = r.begin(); i != r.end(); ++i)
```
It's a good thing we're filling in gaps in our testing...

trws replied:
I agree, though I'm not sure why you're flagging this line? The original worked for all iterators that could be wrapped with the tbb range we were using, which come to find out assumes uses a size_t difference type internally, so it only works if end is reachable by using `++` on begin enough times, which sometimes isn't the case for us currently since we aren't actually providing STL compatible ranges as it turns out.

rhornung67 replied:
Because we didn't have any tests before that combined tbb exec policies with range-stride segments.  Perhaps, my intent wasn't clear  -- it was meant to be kudos to you for figuring out this change was needed and a bit of sarcasm directed at the fact that we weren't covering all our use cases in our tests.

rhornung67 commented on include/RAJA/internal/Iterators.hpp:
```diff
@@ -105,15 +105,16 @@ class numeric_iterator
+  ~numeric_iterator() = default;
+  constexpr numeric_iterator(const numeric_iterator&) = default;
+  constexpr numeric_iterator(numeric_iterator&&) = default;
+  constexpr numeric_iterator& operator=(const numeric_iterator&) = default;
```
@trws with these changes, lines 112, 113 generate errors with the Intel 19 compiler (I didn't try other versions): 'invalid type for defaulted assignment operator'.  Code compiles and runs with clang 9 and gcc 8.  Similar for line 281 below, etc.

Also, new warnings: constexpr non-static member function will not be implicitly 'const' in C++14.   The warnings go away with Intel compiler when 'const' is added before '= default', but gnu and clang then give errors about defaulted copy-operators cannot be declared 'const'

Thoughts on what to do here to satisfy all compilers?

rhornung67 replied:
@ajkunen @willkill07 @mdavis36 if you have ideas about this, please share.

ajkunen replied:
i've run into this before.   If you just say `numeric_iterator& operator=(const numeric_iterator&) = default;` i believe that will automatically be constexpr... I think... but one should check the standard, and also check that the compiler agrees

ajkunen replied:
the most portable thing to do is:  don't use `default`

trws replied:
We can't make them const, but we can parametrize when they are `constexpr` as we have done in other places.  @ajkunen's right that the compiler is supposed to make them `constexpr` automatically if they can be, and explicit `constexpr` is only allowed when it would have been done automatically anyway, so it's basically there to make a compiler complain if it would have made them non-constexpr by mistake.

I can just take `constexpr` out, or make it conditional based on compiler.  This is a case of ICC being buggy, sadly, but clearly we need to work around it one way or another.

rhornung67 replied:
So removing constexpr on the copy and move assignment satisfies clang, gnu, and intel, and all tests pass. Will test cuda and hip next....

If default doesn't give portability, what's the freakin' point?!? Ay carumba!

rhornung67 replied:
@trws I'm running some tests now. I will push some changes shortly. Hopefully, this will get the CI builds to pass.

rhornung67 replied:
Arrrghhh! Spoke too soon. Removing constexpr on the copy and move assignment operators satisfies gnu, clang, nvcc.  Intel generates an internal segmentation violation. 

Now where did I put the contact info for Intel support assistance that it's telling me to report this to....  ❤️

trws replied:
> the most portable thing to do is: don't use `default`

Sadly this is true... we may just have to make all these darned things explicit.  It's unfortunate since we only actually need *one* explicit constructor that isn't even from the default set and lose all the defaulted ones to it.  If it wouldn't be an API break I'd even make it aggregate initialized, but it is. 😢

rhornung67 commented:
Getting closer, but TBB forall-segment tests failing...

trws commented:
I can put up a fix for that, it sounds like folks are wanting option (2) of what I proposed earlier, which should clear the last of that up with an update to RangeStrideSegment.

rhornung67 commented:
Thanks @trws

trws commented:
Follow-up question, if we're doing clamping on RSS should we do it for RangeSegment as well and pull out the throw?

trws commented:
Tried to send this a couple hours ago, but apparently it never posted.  Yay for email.


> On Apr 6, 2020, at 9:53 AM, Rich Hornung <notifications@github.com> wrote:

> @trws It's not clear to me what you mean in #1 by "invalid values". Do you mean when the values given in a range or range-stride segment would result in no loop iterations to execute? To me it makes most sense to follow the semantics of standard C-style for loops. For example,
'for (int i = 2; i < 0; ++i)' would yield no iterations to execute, but the code would still compile and not generate any errors when run (except for unexpected behavior if the user screwed up). However, 'for (int i = 6; i > -1: i -= 2)' would yield iterations with values i = 6, 4, 2, 0.


Note that you put > and < in each of these, we don’t have that from the user, we have ‘!=‘ as the standard interface for iterators, we could make the range behave as though that’s what we’re doing by selecting the comparison operator by stride’s sign though.  We can make our ranges behave this way, but at present if you do TypedRangeSegment(100,1) it will throw an exception treating that as invalid rather than executing zero iterations.

> This would be the same as using TypedRangeSegment(6, -1, -2) with a RAJA forall, which we need to support.

Yes, that works fine as is actually.  The problem is when you do TypedRangeStrideSegment(1, 20, -2) or similar, which would result in a range that can actually reach the end value, but only by invoking undefined behavior through signed type rollover with a distance in the billions (this is what currently happens, size will report 0 but distance(begin(r), end(r)) gives approximately INT_MAX/2 - 10).  I consider that “invalid” because it relies on undefined behavior to have any sane meaning.  In that case we can either say that we clamp it to zero iterations, because the comparison operator we selected from the sign says there are none, or we can throw an error like we currently do for RangeSegment.

> Perhaps, we could add run time bounds checks inserted via our compile-time CMake variable internally, like we do for Views, and report to users that no loop iterations would result?

Sure, we could add that too, but we have to do at least some run-time checking even without that to ensure we don’t do anything actively harmful.  When I tested it I was getting things like 13846537 for the 1,20,-2 range.

rhornung67 commented:
> Follow-up question, if we're doing clamping on RSS should we do it for RangeSegment as well and pull out the throw?

Might as well.

trws commented:
> > Follow-up question, if we're doing clamping on RSS should we do it for RangeSegment as well and pull out the throw?
> 
> Might as well.

Will do then, running tests on the updated strided segment now.

rhornung67 commented:
sounds good. thanks.

trws commented:
Ok, clamped version pushed, done for both range types and broken tests (looking for throw) updated to check for clamping.

rhornung67 commented:
@trws significant progress here.  I can reproduce the hip build issue on corona. The fix is to slap RAJA_HOST_DEVICE before each ctor and dtor in the iterator classes. Of course, then nvcc screams with a bunch of warnings. Does it make sense to create some hip/nvcc-specific macros for this specific case? 

The intel 18 build timed out here. I don't think it got to the point where Intel 19 generates an internal compiler error.  I'm playing around with optimization flags for the compiler now to see if I can get a build to complete....

trws commented:
> @trws significant progress here. I can reproduce the hip build issue on corona. The fix is to slap RAJA_HOST_DEVICE before each ctor and dtor in the iterator classes. Of course, then nvcc screams with a bunch of warnings. Does it make sense to create some hip/nvcc-specific macros for this specific case?

I think we made some for another case like this in CAMP already actually, yeah it’s `CAMP_HIP_HOST_DEVICE` which we could certainly use here if we want to do it that way.  That said, @ajkunen had a point, it may be worth writing them out explicitly to avoid the issue of defaulting them.  I pulled out the two explicit ones on the theory it would be better to just default all of them, but at this point writing out the rest might be worth it.
 
> The intel 18 build timed out here. I don't think it got to the point where Intel 19 generates an internal compiler error. I'm playing around with optimization flags for the compiler now to see if I can get a build to complete....

Ouch... I don’t think we’re using [travis_wait](https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received) in our builds yet, but if we are hitting 10-minute build times per-file with ICC it might be time to add it for that target.  It’s bad enough having to wait for it once, let alone having to restart it repeatedly.

trws commented:
Actually, should I just take care of both of those?  My daughter fell asleep for a nap (🎆 🍾 🎉 🕺  (she doesn’t nap anymore) 😴 👶 ) so I actually have a bit more time to get some technical work done this afternoon.  Might also be good to take a clang time-trace on whatever file timed out, did you happen to notice which it was?

rhornung67 commented:
Go for it.  I've got other stuff I need to work on. 

I'm pretty sure the file is test-forall-indexset-openmp.cpp in test/functional/forall.  Lots of warning spew from the compiler about skipping openmp processing to constrain compile time. Didn't we see this before? It looks like test-forall-indexset-seq.cpp is also taking a lot of time. Not sure what's so hard about these two files. With intel they seem to take a lot longer than the atomic test files used to before they were broken apart.

Thanks for all your help on this.

rhornung67 commented:
@trws FWIW, I was able to get this to build (in a reasonable amount time) and all tests pass with Intel 19 at -O1 optimization level.  -O2 and -O3 generate the internal compiler error I mentioned earlier.

trws commented:
We did have something much like that, it was a bug in forceinline logic at the time I think, but if there is enough going on in there we may just be hitting a similarly nasty degenerate case again.

rhornung67 commented:
Which seems a bit strange since the generated tests for index sets are now simpler than what we had earlier, albeit more comprehensive in terms of what they are testing.  We may be plumbing things through google test machinery in a more complex fashion now however. Dunno?

trws commented:
Running a full build with clang (just tbb and openmp on, it’s what I had configured) with ninja and a little awk-fu on the .ninja_log (just learned about this the other day, better than sliced 🍞 ), here are the build times for the top 10 targets in raja in seconds (this was a parallel build, so there was interference, but it’s on a hefty box):

<img width="538" alt="PNG image" src="https://user-images.githubusercontent.com/660149/78613320-5ba44400-7820-11ea-96cd-2d1dd3e6ce65.png">
(Sorry it’s an image, iPad terminal wasn’t letting me copy text out of tmux)

All three versions of that test I built were hugely expensive, 50 to 160 seconds each.  Something in there is making the compilers unhappy, and a bunch of it seems to be link time, maybe something is getting repeatedly instantiated then thrown away?

rhornung67 commented:
Excessive link times is consistent with what @ajkunen reported a couple of weeks ago in his code. Seems like a good topic to discuss and plan for how to address at the meeting tomorrow.

rhornung67 commented:
@trws changes you committed yesterday now break intel 19 builds. One error is at line 239 at Iterators.hpp (similar error at line 251):

error: a constexpr function cannot have a nonliteral return type "RAJA::Iterators::numeric_iterator<RAJA::Index_type={std::ptrdiff_t={long}}, RAJA::Index_type={std::ptrdiff_t={long}}, RAJA::Index_type={std::ptrdiff_t={long}} *>"^M
    RAJA_HOST_DEVICE friend constexpr numeric_iterator operator+(

rhornung67 commented:
@trws your last commit appears to have resolved intel 19 compiler errors. Thanks! 
Still building at -O1. Haven't gone back to trying higher levels of optimization.

trws commented:
I'm pretty sure both of those are from the same mistake. I included explicit destructors on both iterators, which makes them "non-literal" in C++ standardese causing the ICC error and they were unmarked causing the nvcc error.  Waiting for travis to come back, but this commit should take care of those.

rhornung67 commented:
That last commit rebroke intel and nvcc builds: line 308 in Iterator.hpp -- explicit type is missing, int assumed.

trws commented:
That was a typo that I have no idea how I ended up with.  All seem to be building now, not sure if ICC will finish but that's a separate issue.

rhornung67 commented:
@trws no worries. I didn't intend to hound you. I'll test this out with several compilers and platforms and let you know how it goes.

rhornung67 commented:
Hi @trws So I ran compiles with several compilers and versions and this works everywhere I've tried. Intel 19 compiles and tests pass with '-O1' but barfs higher optimization levels.  Would it be better to crank down the optimization level for the Travis intel build or do something else?

I have a couple more calls to get to today. Tomorrow morning, I will throw something together to make the CI builds use a very reduced number of integral types and add an option to specify at compile time that one wants to run the full spectrum.  I'll try it out locally first and then, if it seems to make sense, I will push it to this PR and see if we can get Intel to complete.  Thanks again for all your help on sorting these issues out.

trws commented:
No problem, glad to hear it seems to be working in a wider area.  Also, 
I remembered later that we already have a “TEST_EXHAUSTIVE” flag 
that I added for the atomic tests a while back, I can throw a check for 
that into the type list real quick and see how much it helps, and remove 
the duplicate `long long` and `long long int` while I’m at it.  Then 
we’ll just have to add in a cron test on travis and probably a 
travis_wait for the cron build and we’ll be in business!

By the way, I meant to put a comment up here, it’s not really critical 
since it’s in test code, but we have a few places where we’ve got 
unprotected (by being in a function or namespace) `using namespace` 
directives in headers in these tests.  It took me a bit to figure out 
why `list` with no qualifier was working in one file until I saw the 
using directive in another header included before it.  Just to keep 
things straight it might be nice to factor those out.

rhornung67 commented:
Thanks, Tom.  I'll work on using fully-qualified names in the tests and remove using directives in the header files.

trws commented:
Oof... intel still takes 30 minutes to deal with this, that's just awful, but at least it works I guess?

rhornung67 commented:
I say declare a moderate victory here and explore other ways to reduce compilation times. Thanks.

rhornung67 commented:
@mdavis36 , @trws , @willkill07 , @davidbeckingsale , @rchen20 This PR has been getting larger. So it doesn't get too big, I think it's good to go as is. Please review at your convenience.

I addressed several reviewer comments that were simple changes. I didn't want to spend time changing things that will eventually be removed if we go ahead with this. I will do more cleanup on the eventual final version.

@mdavis36 and @rchen20 please note that I made some changes to several test files after suggestions from @trws . In particular, removed 'using' directives in header files so all camp usage has full-qualified names, etc. Also, note that @trws put guards into the index type lists for the tests so by default, it will compile tests for a few types and builds and runs faster.

@trws with the limited number of index types, Intel 19 now successfully builds at all optimization levels. It seems that the deep template hierarchies were causing the internal compiler segmentation violation.

trws commented:
I’m good with that plan, and think this is a nice refactor. Thanks Rich!

Sent from my iPhone

On Apr 8, 2020, at 3:43 PM, Rich Hornung <notifications@github.com> wrote:

﻿

@mdavis36<https://github.com/mdavis36> , @trws<https://github.com/trws> , @willkill07<https://github.com/willkill07> , @davidbeckingsale<https://github.com/davidbeckingsale> , @rchen20<https://github.com/rchen20> This PR has been getting larger. So it doesn't get too big, I think it's good to go as is. Please review at your convenience.

I addressed several reviewer comments that were simple changes. I didn't want to spend time changing things that will eventually be removed if we go ahead with this. I will do more cleanup on the eventual final version.

@mdavis36<https://github.com/mdavis36> and @rchen20<https://github.com/rchen20> please note that I made some changes to several test files after suggestions from @trws<https://github.com/trws> . In particular, removed 'using' directives in header files so all camp usage has full-qualified names, etc. Also, note that @trws<https://github.com/trws> put guards into the index type lists for the tests so by default, it will compile tests for a few types and builds and runs faster.

@trws<https://github.com/trws> with the limited number of index types, Intel 19 now successfully builds at all optimization levels. It seems that the deep template hierarchies were causing the internal compiler segmentation violation.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/LLNL/RAJA/pull/790#issuecomment-611231841>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAFBFNJRU4FJXFVJT6UUFU3RLT4ZZANCNFSM4L22K5PQ>.

trws commented:
it's not a time to compile thing, it's more about 1) being explicit, and 2) making it less likely we end up with an ambiguity in names because we've brought every member of the namespace into the global namespace.  It's a good idea to use those directives in implementation files, or functions, but at global scope in a header they leak into every header included after them and everything that includes them, which makes it easy to make a header that's broken in any other include order.
